hare3/remote.go

```diff
@@ -0,0 +1,270 @@
+}
+
+func (h *RemoteHare) Register(sig *signing.EdSigner) {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.log.Info("registered signing key", log.ZShortStringer("id", sig.NodeID()))
+	h.signers[string(sig.NodeID().Bytes())] = sig
+}
+
+func (h *RemoteHare) Start() {
+	current := h.nodeClock.CurrentLayer() + 1
+	enabled := max(current, h.config.EnableLayer, types.GetEffectiveGenesis()+1)
+	disabled := types.LayerID(math.MaxUint32)
+	h.log.Info("started",
+		zap.Inline(&h.config),
+		zap.Uint32("enabled", enabled.Uint32()),
+		zap.Uint32("disabled", disabled.Uint32()),
+	)
+	h.eg.Go(func() error {
+		h.log.Info("remote hare processing starting")
+		for next := enabled; next < disabled; next++ {
+			h.log.Info("remote hare processing layer", zap.Int("next", int(next)))
+			select {
+			case <-h.nodeClock.AwaitLayer(next):
+				h.log.Debug("notified", zap.Uint32("layer", next.Uint32()))
+				h.onLayer(next)
+			case <-h.ctx.Done():
+				h.log.Info("remote hare processing layer - context done")
+				return nil
+			}
+		}
+		return nil
+	})
+}
+
+func (h *RemoteHare) beacon(e types.EpochID) types.Beacon {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	b, ok := h.beacons[e]
+	if !ok {
+		bcn, err := h.svc.Beacon(context.Background(), e)
+		if err != nil {
+			h.log.Error("error getting beacon", zap.Error(err))
+			return types.EmptyBeacon
+		}
+		h.beacons[e] = bcn
+		return bcn
+	}
+
+	return b
+}
+
+func (h *RemoteHare) onLayer(layer types.LayerID) {
+	h.log.Debug("remote hare: on layer", zap.Int("layer", int(layer)))
+	beacon := h.beacon(layer.GetEpoch())
+	if beacon == types.EmptyBeacon {
+		h.log.Debug("no beacon",
+			zap.Uint32("epoch", layer.GetEpoch().Uint32()),
+			zap.Uint32("lid", layer.Uint32()),
+			// zap.Error(err),
+		)
+		return
+	}
+
+	h.mu.Lock()
+	// signer can't join mid session
+	s := &session{
+		lid:     layer,
+		beacon:  beacon,
+		signers: maps.Values(h.signers),
+		vrfs:    make([]*types.HareEligibility, len(h.signers)),
+		proto:   newProtocol(h.config.CommitteeFor(layer)/2 + 1),
+	}
+	h.sessions[layer] = s.proto
+	h.mu.Unlock()
+
+	sessionStart.Inc()
+	// h.tracer.OnStart(layer)
+	h.log.Debug("registered layer", zap.Uint32("lid", layer.Uint32()))
+	h.eg.Go(func() error {
+		if err := h.run(s); err != nil {
+			h.log.Warn("failed",
+				zap.Uint32("lid", layer.Uint32()),
+				zap.Error(err),
+			)
+			exitErrors.Inc()
+			// if terminated successfully it will notify block generator
+			// and it will have to CompleteHare
+		} else {
+			h.log.Debug("terminated",
+				zap.Uint32("lid", layer.Uint32()),
+			)
+		}
+		h.mu.Lock()
+		delete(h.sessions, layer)
+		h.mu.Unlock()
+		sessionTerminated.Inc()
+		return nil
+	})
+}
+
+func (h *RemoteHare) run(session *session) error {
+	var (
+		current = IterRound{Round: preround}
+		start   = time.Now()
+		active  bool
+	)
+	for i := range session.signers {
+		session.vrfs[i] = h.oracle.active(session.signers[i], session.beacon, session.lid, current)
+		active = active || session.vrfs[i] != nil
+	}
```

Comment:
Might be a matter of taste, but...
```suggestion
	for i, signer := range session.signers {
		session.vrfs[i] = h.oracle.active(signer, session.beacon, session.lid, current)
		active = active || session.vrfs[i] != nil
	}
```